File: Java/114.java
__________________________________________________________________________________________________
sample 0 ms submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode prev;
    
    public void flatten(TreeNode root) {
        //base case
        if(root == null){
            return;
        }    
        
        prev = null;
        
        //run it in recursion pre-order
        flattenTree(root); 
    }
    
    //Function to flatten tree with pre-order traversal
    public void flattenTree(TreeNode current){
         if(current == null){
             return;
         }   
        
         //start addiing current to the right of the prev 
        if(prev != null){
            prev.right = current;
            prev.left = null;
        }
        prev = current;
        
        //get all nodes form left --> right and add to the prev
        TreeNode right = current.right;
        flattenTree(current.left);
        flattenTree(right);
    }
}
__________________________________________________________________________________________________
sample 34252 kb submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 04/03/2019
// Runtime: 0 ms, beat 100%

class Solution {
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        }
        flattenHelper(root, new TreeNode[] {null});
    }
    
    private void flattenHelper(TreeNode root, TreeNode[] prev) {
        if (root == null) {
            return;
        } 
        flattenHelper(root.right, prev);
        flattenHelper(root.left, prev);
        root.right = prev[0];
        root.left = null;
        prev[0] = root;
    }
}
__________________________________________________________________________________________________

File: Python3/114.py
__________________________________________________________________________________________________
sample 28 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        st=[root];head1=TreeNode(0)
        if not root:
            return
        head=head1
        
        while st:
            
            x=st.pop()
            head.right=x
            if x.right:
                st.append(x.right)
            if x.left:
                st.append(x.left)
            x.left=None
            head=head.right
        head.right=None
__________________________________________________________________________________________________
sample 13220 kb submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
def get_nodes(root):
    if not root:
        return []
    
    left = get_nodes(root.left)
    right = get_nodes(root.right)
    
    return [root] + left + right 

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        ordered = get_nodes(root)
        for i, node in enumerate(ordered[:-1]):
            node.left = None
            node.right = ordered[i + 1]        
__________________________________________________________________________________________________
